tests/run/lazy-impl.scala

```diff
@@ -0,0 +1,214 @@
+
+    @tailrec def x: A =
+        _x match
+        case current: A =>
+            current
+        case null =>
+            if CAS(_x, null, Evaluating) then
+                var result: A = null
+                try
+                    result = rhs
+                    if result == null then result = NULL // drop if A is non-nullable
+                finally
+                    if !CAS(_x, Evaluating, result) then
+                        val lock = _x.asInstanceOf[Waiting]
+                        CAS(_x, lock, result)
+                        lock.release()
+            x
+        case Evaluating =>
+            CAS(x, Evaluating, new Waiting)
+            x
+        case current: Waiting =>
+            _x = current.awaitRelease()
+            x
+        case NULL => null                                // drop if A is non-nullable
+
+ *  The code makes use of the following runtime class:
+
+    class Waiting:
+        private var done = false
+        def release(): Unit = synchronized:
+            done = true
+            notifyAll()
+
+        def awaitRelease(): Unit = synchronized:
+            while !done do wait()
+
+ *  Note: The code assumes that the getter result type `A` is disjoint from the type
+ *  of `Evaluating` and the `Waiting` class. If this is not the case (e.g. `A` is AnyRef),
+ *  then the conditions in the match have to be re-ordered so that case `_x: A` becomes
+ *  the final default case.
+ *
+ *  Cost analysis:
+ *
+ *    - 2 CAS on contention-free initialization
+ *    - 0 or 1 CAS on first read in thread other than initializer thread, depending on
+ *      whether cache has updated
+ *    - no synchronization operations on reads after the first one
+ *    - If there is contention, we see in addition
+ *       - for the initializing thread: another CAS and a synchronized notifyAll
+ *       - for a reading thread: 0 or 1 CAS and a synchronized wait
+ *
+ *  Code sizes for getter:
+ *
+ *   this scheme, if nulls are excluded in type: 86 bytes
+ *   current Dotty scheme: 125 bytes
+ *   Scala 2 scheme: 39 bytes
+ *
+ *  Advantages of the scheme:
+ *
+ *   - no slot other than the field itself is needed
+ *   - no locks are shared among lazy val initializations and between lazy val initializations
+ *     and normal code
+ *   - no deadlocks (other than those inherent in user code)
+ *   - synchronized code is executed only if there is contention
+ *   - simpler than current Dotty scheme
+ *
+ *  Disadvantages:
+ *
+ *   - lazy vals of primitive types are boxed
+ */
+import sun.misc.Unsafe._
+
+class C {
+  def init(name: String) = {
+    Thread.sleep(10)
+    println(s"initialize $name"); "result"
+  }
+
+  @volatile private[this] var _x: AnyRef = _
+
+  // Expansion of:  lazy val x: String = init("x")
+
+  def x: String = {
+    val current = _x
+    if (current.isInstanceOf[String])
+      current.asInstanceOf[String]
+    else
+      x$lzy
+  }
+
+  def x$lzy: String = {
```

Comment:
Currently `Tailrec` runs before `LazyVals`. We need to write the loop manually:
```scala
def x$lzy: String = {
  while (<EmptyTree>) {
    val current = _x
    if (current.isInstanceOf[String])
      return current.asInstanceOf[String]
    else {
      val offset = C.x_offset
      if (current == null) {
        if (LazyRuntime.isUnitialized(this, offset)) {
          try {
            val result = init("x")
            LazyRuntime.initialize(this, offset, result)
            return result
          catch {
            case ex: Throwable =>
              LazyRuntime.initialize(this, offset, null)
              throw ex
          }
        }
      }
      else
        LazyRuntime.awaitInitialized(this, offset, current)
    }
  }
}
```
Note the `while(<EmptyTree>)`.  We special case it in the compiler. It means infinite loop and types to `Nothing`: https://github.com/lampepfl/dotty/blob/963719ed2679f3d4c8188ec068e53941b181ef76/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala#L529-L530

Reply:
Good to know. That's trap one could fall into easily.

Reply:
Might want to rethrow the original exception in case the second initialize fails.

Reply:
I don't think the second initialize can fail. There's no user code exectuted for it.

Reply:
@odersky Good point. I guess in the cases it fails it is something like a SoE or similar VME.